File: core/tests/test_passives_loss_thermal.cpp
// Capacitor + Resistor + Inductor loss / thermal — Catch2 tests
// (Phase 3 of inverter-bridge-losses, Pulsim 0.10.0a9/a10/a11).
//
// Each passive now exposes the same loss-accumulator + thermal
// pattern as the switching devices:
//     Capacitor : P = I_cap² · ESR(T_j)
//     Resistor  : P = V²/R(T_j)  (= I²·R)
//     Inductor  : P = I² · DCR(T_j)
//
// Backward-compat is verified by the legacy-mode tests (R_th_ja == 0
// → accumulator is a no-op).

#include <catch2/catch_test_macros.hpp>
#include <catch2/catch_approx.hpp>

#include "pulsim/v1/simulation.hpp"

#include <cmath>

using namespace pulsim::v1;
using Catch::Approx;

namespace {

SimulationOptions make_opts(Real tstop, Real dt) {
    SimulationOptions opts;
    opts.tstart = 0.0;
    opts.tstop = tstop;
    opts.dt = dt;
    opts.dt_min = 1e-9;
    opts.dt_max = dt;
    opts.adaptive_timestep = false;
    opts.enable_bdf_order_control = false;
    return opts;
}

}  // namespace

// ---------------------------------------------------------------------------
// Capacitor ESR loss
// ---------------------------------------------------------------------------
TEST_CASE("Capacitor: legacy mode (R_th_ja=0) leaves loss accumulator at 0",
          "[passives_loss][capacitor][regression]") {
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd  = Circuit::ground();
    circuit.add_voltage_source("Vs", n_in, gnd, 5.0);
    circuit.add_capacitor("C1", n_in, gnd, 1e-6);

    auto opts = make_opts(1e-3, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    CHECK(circuit.capacitor_average_power("C1") == Approx(0.0));
    CHECK(circuit.capacitor_total_energy("C1") == Approx(0.0));
}

TEST_CASE("Capacitor: ESR loss for sinusoidal ripple current",
          "[passives_loss][capacitor][regression]") {
    // Drive a known ripple current through the cap and verify the
    // accumulated P_avg matches I_rms² · ESR.
    Circuit circuit;
    const Index a = circuit.add_node("a");
    const Index gnd = Circuit::ground();

    // 1 A peak (≈ 0.707 A rms) sinusoidal current source @ 1 kHz.
    SineParams sp{};
    sp.amplitude = 1.0;
    sp.frequency = 1000.0;
    sp.offset = 0.0;
    sp.phase = 0.0;
    // Note: there's no add_sine_current_source; build current via a
    // simpler approach — a fixed DC current through a series resistor
    // with the cap in parallel-shunt won't excite ESR meaningfully.
    // Instead use the high-frequency voltage swing across the cap.
    sp.amplitude = 100.0;
    circuit.add_sine_voltage_source("Vac", a, gnd, sp);

    Capacitor::Params cp{};
    cp.capacitance = 1e-6;
    cp.ESR         = 0.1;        // 100 mΩ
    cp.ESR_tc      = 0.0;         // no thermal feedback for clean math
    cp.R_th_ja     = 5.0;
    cp.T_amb       = 25.0;
    circuit.add_capacitor("C1", a, gnd, cp);

    auto opts = make_opts(20e-3, 5e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    // I_cap = C · dV/dt = C · ω · V_peak · cos(ωt)
    //   → I_rms = C · ω · V_peak / √2
    //   → P = I_rms² · ESR
    const Real omega = 2.0 * 3.14159265358979323846 * 1000.0;
    const Real I_rms = cp.capacitance * omega * 100.0 / std::sqrt(2.0);
    const Real P_expected = I_rms * I_rms * cp.ESR;

    const Real p_avg = circuit.capacitor_average_power("C1");
    INFO("Expected steady-state I_rms = " << I_rms << " A → P ≈ "
         << P_expected << " W (analytical, ignoring trap. startup). "
         "Measured P_avg = " << p_avg << " W. The first few timesteps "
         "have trapezoidal-companion startup ringing that inflates I — "
         "the order of magnitude check is what matters here.");
    CHECK(p_avg > 0.0);
    // Order-of-magnitude check: within 5× of analytical (trap startup
    // ringing on first cycle can inflate I_rms over the full window).
    CHECK(p_avg < 5.0 * P_expected);
    CHECK(p_avg > 0.2 * P_expected);

    // T_j_derived = T_amb + P_avg · R_th_ja
    const Real t_j = circuit.capacitor_steady_state_junction_temperature("C1");
    CHECK(t_j == Approx(cp.T_amb + p_avg * cp.R_th_ja).epsilon(0.001));
    CHECK(t_j > cp.T_amb);
}

// ---------------------------------------------------------------------------
// Resistor I²R loss
// ---------------------------------------------------------------------------
TEST_CASE("Resistor: legacy mode (R_th_ja=0) leaves loss accumulator at 0",
          "[passives_loss][resistor][regression]") {
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd  = Circuit::ground();
    circuit.add_voltage_source("Vs", n_in, gnd, 10.0);
    circuit.add_resistor("R1", n_in, gnd, 100.0);

    auto opts = make_opts(1e-3, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    CHECK(circuit.resistor_average_power("R1") == Approx(0.0));
    CHECK(circuit.resistor_total_energy("R1") == Approx(0.0));
}

TEST_CASE("Resistor: P = V²/R for DC operation",
          "[passives_loss][resistor][regression]") {
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd  = Circuit::ground();

    Resistor::Params rp{};
    rp.resistance = 10.0;
    rp.TCR        = 0.0;            // no thermal feedback for clean math
    rp.R_th_ja    = 2.0;
    rp.T_amb      = 25.0;
    circuit.add_resistor("R1", n_in, gnd, rp);
    circuit.add_voltage_source("Vs", n_in, gnd, 10.0);

    auto opts = make_opts(10e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    // P_expected = V²/R = 100/10 = 10 W.
    const Real P_expected = 100.0 / rp.resistance;
    const Real p_avg = circuit.resistor_average_power("R1");
    INFO("Expected P = " << P_expected << " W. Measured = " << p_avg << " W");
    CHECK(p_avg == Approx(P_expected).epsilon(0.05));

    const Real t_j = circuit.resistor_steady_state_junction_temperature("R1");
    CHECK(t_j == Approx(rp.T_amb + p_avg * rp.R_th_ja).epsilon(0.001));
    CHECK(t_j > rp.T_amb);
}

TEST_CASE("Resistor: TCR feedback raises loss when T_j increases",
          "[passives_loss][resistor][regression]") {
    // The Phase-3 loss accumulator follows the powerStage convention:
    // the SIMULATOR sees R_nominal (cold) and produces a current I; the
    // accumulator integrates I² · R(T_j) using the hot resistance from
    // the TCR formula. So P grows with T_j (more dissipation at hot).
    // If you also want the hot R to reduce current in the simulation
    // itself, set_resistance(R(T_j)) before re-running.
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd  = Circuit::ground();

    Resistor::Params rp{};
    rp.resistance = 10.0;
    rp.TCR        = 5e-3;          // 0.5 %/K (typical metal-film)
    rp.T_ref      = 25.0;
    rp.R_th_ja    = 1.0;
    rp.T_amb      = 25.0;
    circuit.add_resistor("R1", n_in, gnd, rp);
    circuit.add_voltage_source("Vs", n_in, gnd, 10.0);

    auto opts = make_opts(10e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();

    // Pass 1: T_j = T_amb = 25 °C. I = 1 A, P = I²·R(25) = 10 W.
    {
        REQUIRE(Simulator(circuit, opts).run_transient().success);
    }
    const Real p_25 = circuit.resistor_average_power("R1");

    // Pass 2: T_j = 125 °C → R(125) = 10·(1 + 0.005·100) = 15 Ω.
    // I still = 1 A (simulator uses R_nom). P = I²·R(125) = 15 W.
    circuit.set_resistor_T_j("R1", 125.0);
    circuit.reset_resistor_loss("R1");
    {
        REQUIRE(Simulator(circuit, opts).run_transient().success);
    }
    const Real p_125 = circuit.resistor_average_power("R1");

    INFO("P @ T_j=25 °C: " << p_25 << " W; P @ T_j=125 °C: " << p_125 << " W");
    CHECK(p_125 > p_25);                       // hot dissipation higher
    CHECK((p_125 - p_25) / p_25 > 0.30);       // ≥ 30 % rise (≈ 50 % nominal)
}

// ---------------------------------------------------------------------------
// Inductor DCR loss
// ---------------------------------------------------------------------------
TEST_CASE("Inductor: legacy mode (R_th_ja=0) leaves loss accumulator at 0",
          "[passives_loss][inductor][regression]") {
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd  = Circuit::ground();
    circuit.add_voltage_source("Vs", n_in, gnd, 5.0);
    circuit.add_inductor("L1", n_in, gnd, 1e-3);

    auto opts = make_opts(1e-3, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    CHECK(circuit.inductor_average_power("L1") == Approx(0.0));
    CHECK(circuit.inductor_total_energy("L1") == Approx(0.0));
}

TEST_CASE("Inductor: P = I² · DCR for DC current",
          "[passives_loss][inductor][regression]") {
    // Series chain V → L → R_load → GND. At steady state, the inductor
    // looks like a short (DC) so I = V/(R_load + DCR) — but the DCR
    // is NOT in the stamping, only in the loss accumulator. So I in
    // the simulation = V/R_load, and accumulator computes I²·DCR.
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index n_mid = circuit.add_node("mid");
    const Index gnd  = Circuit::ground();

    Inductor::Params lp{};
    lp.inductance = 1e-3;
    lp.DCR        = 0.05;        // 50 mΩ winding
    lp.DCR_tc     = 0.0;          // no thermal feedback
    lp.R_th_ja    = 3.0;
    lp.T_amb      = 25.0;
    circuit.add_inductor("L1", n_in, n_mid, lp);

    const Real R_load = 1.0;
    circuit.add_resistor("Rload", n_mid, gnd, R_load);
    circuit.add_voltage_source("Vs", n_in, gnd, 5.0);

    auto opts = make_opts(50e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    REQUIRE(Simulator(circuit, opts).run_transient().success);

    // I_steady = V/R_load = 5 A.  P = I²·DCR = 25·0.05 = 1.25 W.
    const Real I_expected = 5.0 / R_load;
    const Real P_expected = I_expected * I_expected * lp.DCR;
    const Real p_avg = circuit.inductor_average_power("L1");
    INFO("Expected I = " << I_expected << " A, P = " << P_expected
         << " W. Measured P_avg = " << p_avg << " W");
    CHECK(p_avg == Approx(P_expected).epsilon(0.10));
    CHECK(p_avg > 0.0);

    const Real t_j = circuit.inductor_steady_state_junction_temperature("L1");
    CHECK(t_j == Approx(lp.T_amb + p_avg * lp.R_th_ja).epsilon(0.001));
}

TEST_CASE("Passives: T_j accessors return NaN for unknown name",
          "[passives_loss][api]") {
    Circuit circuit;
    CHECK(std::isnan(circuit.capacitor_average_power("none")));
    CHECK(std::isnan(circuit.capacitor_junction_temperature("none")));
    CHECK(std::isnan(circuit.resistor_average_power("none")));
    CHECK(std::isnan(circuit.resistor_junction_temperature("none")));
    CHECK(std::isnan(circuit.inductor_average_power("none")));
    CHECK(std::isnan(circuit.inductor_junction_temperature("none")));
    // setters and reset are silent no-ops on unknown names.
    circuit.set_capacitor_T_j("none", 80.0);
    circuit.reset_capacitor_loss("none");
    circuit.set_resistor_T_j("none", 80.0);
    circuit.reset_resistor_loss("none");
    circuit.set_inductor_T_j("none", 80.0);
    circuit.reset_inductor_loss("none");
    CHECK(true);
}

File: legacy/v1/python/tests/test_passives_loss_thermal.py
"""Smoke tests for Capacitor / Resistor / Inductor loss + thermal
(Phase 3 of inverter-bridge-losses, Pulsim 0.10.0a9).

Verifies the pybind boundary for ResistorParams / CapacitorParams /
InductorParams plus the matching Circuit accessors. Mirrors the
Catch2 suite in `core/tests/test_passives_loss_thermal.cpp`.
"""
from __future__ import annotations

import math

import pytest

import pulsim as ps


def _run(circuit, tstop=10e-3, dt=100e-6):
    opts = ps.SimulationOptions()
    opts.tstart = 0.0
    opts.tstop = tstop
    opts.dt = dt
    opts.dt_min = 1e-9
    opts.dt_max = dt
    opts.adaptive_timestep = False
    opts.enable_bdf_order_control = False
    opts.newton_options.num_nodes = circuit.num_nodes()
    opts.newton_options.num_branches = circuit.num_branches()
    return ps.Simulator(circuit, opts).run_transient()


class TestPassivesBindings:
    """Verify Params classes + Circuit accessors are exposed."""

    def test_resistor_params_exposed(self):
        p = ps.ResistorParams()
        p.resistance = 100.0
        p.TCR = 4e-3
        p.R_th_ja = 50.0
        p.T_amb = 25.0
        assert p.resistance == pytest.approx(100.0)
        assert p.TCR == pytest.approx(4e-3)
        assert p.R_th_ja == pytest.approx(50.0)

    def test_capacitor_params_exposed(self):
        p = ps.CapacitorParams()
        p.capacitance = 330e-6
        p.ESR = 0.5
        p.ESR_tc = 0.01
        p.R_th_ja = 8.0
        assert p.capacitance == pytest.approx(330e-6)
        assert p.ESR == pytest.approx(0.5)
        assert p.R_th_ja == pytest.approx(8.0)

    def test_inductor_params_exposed(self):
        p = ps.InductorParams()
        p.inductance = 200e-6
        p.DCR = 0.05
        p.R_th_ja = 12.0
        assert p.inductance == pytest.approx(200e-6)
        assert p.DCR == pytest.approx(0.05)

    def test_circuit_methods_present(self):
        c = ps.Circuit()
        for prefix in ("capacitor", "resistor", "inductor"):
            for suffix in (
                "_average_power", "_peak_power", "_total_energy",
                "_junction_temperature",
                "_steady_state_junction_temperature", "_last_current",
            ):
                assert hasattr(c, prefix + suffix), f"missing {prefix + suffix}"
        for prefix in ("capacitor", "resistor", "inductor"):
            assert hasattr(c, f"set_{prefix}_T_j")
            assert hasattr(c, f"reset_{prefix}_loss")


class TestPassivesBehavior:
    """End-to-end Python validation of the loss accumulator."""

    def test_resistor_loss_tracked_without_thermal_feedback(self):
        # Unified pipeline (Pulsim 0.10.0a12+): the per-device loss
        # accumulator runs always. `R_th_ja == 0` only disables T_j
        # feedback (the resistance is taken as the static value).
        c = ps.Circuit()
        n = c.add_node("n")
        c.add_voltage_source("Vs", n, ps.Circuit.ground(), 10.0)
        c.add_resistor("R1", n, ps.Circuit.ground(), 100.0)
        r = _run(c)
        assert r.success
        # P = V²/R = 100/100 = 1 W (no thermal feedback).
        assert c.resistor_average_power("R1") == pytest.approx(1.0, rel=0.05)

    def test_resistor_with_thermal_dissipates_VV_over_R(self):
        c = ps.Circuit()
        n = c.add_node("n")
        c.add_voltage_source("Vs", n, ps.Circuit.ground(), 10.0)
        rp = ps.ResistorParams()
        rp.resistance = 10.0
        rp.TCR = 0.0
        rp.R_th_ja = 2.0
        rp.T_amb = 25.0
        c.add_resistor("R1", n, ps.Circuit.ground(), rp)
        r = _run(c)
        assert r.success
        # P = V²/R = 100/10 = 10 W
        assert c.resistor_average_power("R1") == pytest.approx(10.0, rel=0.05)
        # T_j = T_amb + P · R_th_ja = 25 + 10·2 = 45 °C
        t_j = c.resistor_steady_state_junction_temperature("R1")
        assert t_j == pytest.approx(45.0, rel=0.05)

    def test_inductor_DCR_loss(self):
        # V → L → R_load → GND. Steady-state I = V/R_load.
        c = ps.Circuit()
        a = c.add_node("a")
        b = c.add_node("b")
        c.add_voltage_source("Vs", a, ps.Circuit.ground(), 5.0)
        lp = ps.InductorParams()
        lp.inductance = 1e-3
        lp.DCR = 0.05
        lp.DCR_tc = 0.0
        lp.R_th_ja = 3.0
        lp.T_amb = 25.0
        c.add_inductor("L1", a, b, lp)
        c.add_resistor("Rload", b, ps.Circuit.ground(), 1.0)
        r = _run(c, tstop=50e-3, dt=100e-6)
        assert r.success
        # I_steady = 5/1 = 5 A. P = 25·0.05 = 1.25 W.
        assert c.inductor_average_power("L1") == pytest.approx(1.25, rel=0.10)

    def test_capacitor_legacy_mode_zero_loss(self):
        c = ps.Circuit()
        n = c.add_node("n")
        c.add_voltage_source("Vs", n, ps.Circuit.ground(), 5.0)
        c.add_capacitor("C1", n, ps.Circuit.ground(), 1e-6)
        r = _run(c)
        assert r.success
        assert c.capacitor_average_power("C1") == pytest.approx(0.0)


class TestPowerStageCapMatch:
    """A focused sanity check that mirrors the user's notebook setup:
    330 µF/400 V bulk cap on a single AC source. Verifies the
    accumulator produces a finite, non-trivial dissipation."""

    def test_330uf_cap_with_ESR(self):
        c = ps.Circuit()
        ac = c.add_node("ac")
        gnd = ps.Circuit.ground()
        # 100 V_rms / 60 Hz AC source (matches the inverter_550W
        # operating point your notebook uses).
        sine = ps.SineParams()
        sine.amplitude = 100.0 * math.sqrt(2.0)
        sine.frequency = 60.0
        sine.offset = 0.0
        sine.phase = 0.0
        c.add_sine_voltage_source("Vac", ac, gnd, sine)

        cp = ps.CapacitorParams()
        cp.capacitance = 330e-6
        cp.ESR = 0.05      # typical generic 330 µF/400 V AlElCap ESR
        cp.ESR_tc = 0.0
        cp.R_th_ja = 8.0
        cp.T_amb = 60.0
        c.add_capacitor("Cbulk", ac, gnd, cp)

        r = _run(c, tstop=50e-3, dt=20e-6)
        assert r.success
        p_avg = c.capacitor_average_power("Cbulk")
        t_j   = c.capacitor_steady_state_junction_temperature("Cbulk")
        assert math.isfinite(p_avg)
        assert p_avg > 0.0
        assert math.isfinite(t_j)
        assert t_j > cp.T_amb
